File: benchmarks/core-comparison/programs/welford.c
#include <math.h>
#include <stdio.h>

typedef struct {
    double count;
    double mean;
    double m2;
} State;

static State welford_step(State state, double value) {
    const double count = state.count + 1.0;
    const double delta = value - state.mean;
    const double mean = state.mean + delta / count;
    const double delta2 = value - mean;
    const State next = {count, mean, state.m2 + delta * delta2};
    return next;
}

static double population_stddev(const volatile double *data, int length) {
    State state = {0.0, 0.0, 0.0};
    for (int index = 0; index < length; ++index) {
        state = welford_step(state, data[index]);
    }
    return sqrt(state.m2 / state.count);
}

int main(void) {
    volatile double values[] = { {{VALUES}} };
    printf("%.17g\n", population_stddev(values, {{COUNT}}));
    return 0;
}

File: benchmarks/core-comparison/programs/welford.rs
#[derive(Clone, Copy)]
struct State {
    count: f64,
    mean: f64,
    m2: f64,
}

fn welford_step(state: State, value: f64) -> State {
    let count = state.count + 1.0;
    let delta = value - state.mean;
    let mean = state.mean + delta / count;
    let delta2 = value - mean;
    State { count, mean, m2: state.m2 + delta * delta2 }
}

fn population_stddev(data: &[f64]) -> f64 {
    let mut state = State { count: 0.0, mean: 0.0, m2: 0.0 };
    for &value in data {
        state = welford_step(state, value);
    }
    (state.m2 / state.count).sqrt()
}

fn main() {
    let values = std::hint::black_box(vec![{{VALUES}}]);
    println!("{:.17}", population_stddev(&values));
}
